modules/rostests/apitests/shlwapi/AssocQueryString.c

```diff
@@ -0,0 +1,118 @@
+/*
+ * PROJECT:     ReactOS api tests
+ * LICENSE:     LGPL-2.0-or-later (https://spdx.org/licenses/LGPL-2.0-or-later)
+ * PURPOSE:     Test for AssocQueryStringA/W
+ * COPYRIGHT:   Copyright 2024 Katayama Hirofumi MZ (katayama.hirofumi.mz@gmail.com)
+ */
+
+#include <apitest.h>
+#include <shlwapi.h>
+#include <stdio.h>
+#include <versionhelpers.h>
+
+typedef HRESULT (WINAPI *FN_AssocQueryStringA)(
+    ASSOCF cfFlags,
+    ASSOCSTR str,
+    LPCSTR pszAssoc,
+    LPCSTR pszExtra,
+    LPSTR pszOut,
+    DWORD *pcchOut);
+typedef HRESULT (WINAPI *FN_AssocQueryStringW)(
+    ASSOCF cfFlags,
+    ASSOCSTR str,
+    LPCWSTR pszAssoc,
+    LPCWSTR pszExtra,
+    LPWSTR pszOut,
+    DWORD *pcchOut);
+
+static HINSTANCE s_hSHLWAPI = NULL;
+static FN_AssocQueryStringA s_fnAssocQueryStringA = NULL;
+static FN_AssocQueryStringW s_fnAssocQueryStringW = NULL;
+static CHAR  s_szTextFileA[MAX_PATH] =  "";
+static WCHAR s_szTextFileW[MAX_PATH] = L"";
+
+static void TEST_Start(void)
+{
+    ExpandEnvironmentStringsA( "%APPDATA%\\This is a test.txt", s_szTextFileA, _countof(s_szTextFileA));
+    ExpandEnvironmentStringsW(L"%APPDATA%\\This is a test.txt", s_szTextFileW, _countof(s_szTextFileW));
+    fclose(_wfopen(s_szTextFileW, L"w"));
+    trace("%s\n", wine_dbgstr_a(s_szTextFileA));
+    trace("%s\n", wine_dbgstr_w(s_szTextFileW));
+}
+
+static void TEST_End(void)
+{
+    DeleteFileW(s_szTextFileW);
+}
+
+/*
+ * "shlwapi_winetest assoc" already has many tests.
+ * We just do additional tests in here.
+ */
+
+static void TEST_AssocQueryStringA(void)
+{
+    CHAR szPath[MAX_PATH], szAnswer[MAX_PATH];
+    HRESULT hr;
+    DWORD cch;
+
+    ExpandEnvironmentStringsA("%WINDIR%\\system32\\notepad.exe", szAnswer, _countof(szAnswer));
+
+    lstrcpynA(szPath, s_szTextFileA, _countof(szPath));
+    cch = _countof(szPath);
+    hr = AssocQueryStringA(0, ASSOCSTR_EXECUTABLE, szPath, NULL, szPath, &cch);
+
+    if (IsWindowsVistaOrGreater())
+    {
+        ok_long(hr, S_OK);
+        ok(lstrcmpiA(szPath, szAnswer) == 0, "%s vs %s\n", wine_dbgstr_a(szPath), wine_dbgstr_a(szAnswer));
+        ok_int(cch, lstrlenA(szAnswer) + 1);
+    }
+    else
+    {
+        ok_long(hr, HRESULT_FROM_WIN32(ERROR_FILE_NOT_FOUND));
```

Comment:
Is this different error understandable? (seeing that this doesn't happen with the unicode version tested below)

Reply:
It seems its Ansi version is imperfect in non-Vista+.